File: tools/session_journal.py
"""Session journal JSON schema (issue #47) — validate exports from AC Copilot Trainer."""

from __future__ import annotations

from typing import Any

JOURNAL_SCHEMA_VERSION = 1

REQUIRED_TOP_LEVEL_KEYS = frozenset(
    {
        "schema_version",
        "exported_at",
        "app_version_ui",
        "session_key",
        "car",
        "track",
        "conditions",
        "summary",
        "lap_history",
        "corners_last_lap",
        "coaching_hints_last",
        "llm_debrief",
    }
)


def validate_session_journal(obj: Any) -> list[str]:
    """Return a list of human-readable validation errors; empty means OK."""
    errors: list[str] = []
    if not isinstance(obj, dict):
        return ["root must be a JSON object"]
    missing = REQUIRED_TOP_LEVEL_KEYS - obj.keys()
    if missing:
        errors.append(f"missing keys: {sorted(missing)}")
    extra = obj.keys() - REQUIRED_TOP_LEVEL_KEYS
    if extra:
        errors.append(f"unknown keys: {sorted(extra)}")

    sv = obj.get("schema_version")
    if sv is not None and sv != JOURNAL_SCHEMA_VERSION:
        errors.append(f"schema_version must be {JOURNAL_SCHEMA_VERSION}, got {sv!r}")

    exported = obj.get("exported_at")
    if exported is not None:
        if not isinstance(exported, str) or "T" not in exported or not exported.endswith("Z"):
            errors.append("exported_at must be an ISO-8601 UTC string ending with Z")

    sk = obj.get("session_key")
    if sk is not None and (not isinstance(sk, str) or not sk):
        errors.append("session_key must be a non-empty string")

    for label, key in (("car", "car"), ("track", "track")):
        sub = obj.get(key)
        if sub is not None:
            if not isinstance(sub, dict):
                errors.append(f"{label} must be an object")
            elif "id" not in sub:
                errors.append(f"{label} must contain id")

    cond = obj.get("conditions")
    if cond is not None:
        if not isinstance(cond, dict):
            errors.append("conditions must be an object")
        elif "track_grip" in cond and cond["track_grip"] is not None:
            if not isinstance(cond["track_grip"], (int, float)):
                errors.append("conditions.track_grip must be a number or null")

    summary = obj.get("summary")
    if summary is not None:
        if not isinstance(summary, dict):
            errors.append("summary must be an object")
        else:
            for k in ("laps_completed", "best_lap_ms", "last_lap_ms", "avg_lap_ms"):
                if (
                    k in summary
                    and summary[k] is not None
                    and not isinstance(summary[k], (int, float))
                ):
                    errors.append(f"summary.{k} must be a number or null")

    lap_hist = obj.get("lap_history")
    if lap_hist is not None:
        if not isinstance(lap_hist, list):
            errors.append("lap_history must be an array")
        else:
            for i, row in enumerate(lap_hist):
                if not isinstance(row, dict):
                    errors.append(f"lap_history[{i}] must be an object")
                    break
                if "lap_ms" in row and row["lap_ms"] is not None:
                    if not isinstance(row["lap_ms"], (int, float)):
                        errors.append(f"lap_history[{i}].lap_ms must be a number")

    corners = obj.get("corners_last_lap")
    if corners is not None and not isinstance(corners, list):
        errors.append("corners_last_lap must be an array")

    hints = obj.get("coaching_hints_last")
    if hints is not None:
        if not isinstance(hints, list):
            errors.append("coaching_hints_last must be an array")
        else:
            for i, h in enumerate(hints):
                if isinstance(h, dict):
                    if "text" not in h:
                        errors.append(f"coaching_hints_last[{i}] missing text")
                elif not isinstance(h, str):
                    errors.append(f"coaching_hints_last[{i}] must be object or string")

    return errors


def sample_valid_session_journal() -> dict[str, Any]:
    """Fixture aligned with Lua `session_journal.buildRecord` output."""
    return {
        "schema_version": JOURNAL_SCHEMA_VERSION,
        "exported_at": "2026-04-03T12:00:00Z",
        "app_version_ui": "v0.4.2",
        "session_key": "ks_lamborghini_huracan_gt3__ks_nordschleife",
        "car": {"id": "ks_lamborghini_huracan_gt3"},
        "track": {"id": "ks_nordschleife"},
        "conditions": {"track_grip": 1.0},
        "summary": {
            "laps_completed": 3,
            "best_lap_ms": 420000,
            "last_lap_ms": 425000,
            "avg_lap_ms": 428000,
        },
        "lap_history": [
            {"lap_ms": 430000, "corner_count": 12},
            {"lap_ms": 425000, "corner_count": 12},
            {"lap_ms": 420000, "corner_count": 12},
        ],
        "corners_last_lap": [
            {
                "label": "T1",
                "entrySpeed": 180.0,
                "minSpeed": 95.0,
                "exitSpeed": 140.0,
                "brakePointSpline": 0.12,
                "trailBrakeRatio": 0.4,
            }
        ],
        "coaching_hints_last": [
            {"kind": "brake", "text": "T1: try braking slightly earlier"},
        ],
        "llm_debrief": None,
    }

File: src/ac_copilot_trainer/modules/session_journal.lua
-- Session training journal (issue #9 Part H, milestone #47): JSON export under ScriptConfig.

local M = {}

local persistence = require("persistence")
local ch = require("csp_helpers")

local SCHEMA_VERSION = 1

local function safeSimField(sim, key)
  if not sim then
    return nil
  end
  local ok, v = pcall(function()
    return sim[key]
  end)
  if ok then
    return v
  end
  return nil
end

--- Match `persistence.sessionKey` fallbacks: globals first, then `car` / `sim` (pcall per field; C-structs throw).
local function resolveCarId(car)
  local g = ch.carIdRawFromGlobals()
  if g and g ~= "" then
    return g
  end
  if not car then
    return "unknown"
  end
  for _, key in ipairs({ "id", "name", "driverName" }) do
    local ok, v = pcall(function()
      return car[key]
    end)
    if ok and v ~= nil and tostring(v) ~= "" then
      return tostring(v)
    end
  end
  return "unknown"
end

local function resolveTrackId(sim)
  local g = ch.trackIdRawFromGlobals()
  if g and g ~= "" then
    return g
  end
  if not sim then
    return "unknown"
  end
  for _, key in ipairs({ "trackName", "track", "trackConfiguration" }) do
    local ok, v = pcall(function()
      return sim[key]
    end)
    if ok and v ~= nil and tostring(v) ~= "" then
      return tostring(v)
    end
  end
  return "unknown"
end

local function logJournal(msg)
  if ac and type(ac.log) == "function" then
    ac.log("[COPILOT] session_journal: " .. msg)
  end
end

local function isoUtcNow()
  return os.date("!%Y-%m-%dT%H:%M:%SZ")
end

local function fileTimestampUtc()
  return os.date("!%Y%m%d_%H%M%S")
end

local function journalDir()
  return persistence.dataDir() .. "/journal"
end

local function indexPath()
  return journalDir() .. "/journal_index.jsonl"
end

local function serializeCoaching(lines)
  if not lines or type(lines) ~= "table" or #lines == 0 then
    return {}
  end
  local out = {}
  for i = 1, #lines do
    local e = lines[i]
    if type(e) == "table" and type(e.text) == "string" then
      out[#out + 1] = {
        kind = type(e.kind) == "string" and e.kind or "general",
        text = e.text,
      }
    elseif type(e) == "string" then
      out[#out + 1] = { kind = "general", text = e }
    end
  end
  return out
end

local function simplifyCorners(corners)
  if not corners or type(corners) ~= "table" then
    return {}
  end
  local out = {}
  for i = 1, #corners do
    local c = corners[i]
    if type(c) == "table" then
      out[#out + 1] = {
        label = c.label,
        entrySpeed = tonumber(c.entrySpeed),
        minSpeed = tonumber(c.minSpeed),
        exitSpeed = tonumber(c.exitSpeed),
        brakePointSpline = tonumber(c.brakePointSpline),
        trailBrakeRatio = tonumber(c.trailBrakeRatio),
      }
    end
  end
  return out
end

---@return table[], number|nil, number|nil, number|nil
local function lapHistorySummaries(hist)
  if not hist or type(hist) ~= "table" then
    return {}, nil, nil, nil
  end
  local laps = {}
  local sum, n = 0, 0
  local bestMs, lastMs = nil, nil
  for i = 1, #hist do
    local lap = hist[i]
    if type(lap) == "table" then
      local ms = tonumber(lap.lapMs)
      if ms and ms > 0 then
        laps[#laps + 1] = {
          lap_ms = ms,
          corner_count = (lap.corners and type(lap.corners) == "table") and #lap.corners or 0,
        }
        sum = sum + ms
        n = n + 1
        lastMs = ms
        if bestMs == nil or ms < bestMs then
          bestMs = ms
        end
      end
    end
  end
  local avgMs = (n > 0) and math.floor(sum / n + 0.5) or nil
  return laps, bestMs, lastMs, avgMs
end

--- Build the on-disk record. Caller supplies `state` snapshot from the driving session.
---@param car ac.StateCar|nil
---@param sim ac.StateSim|nil
---@param state table
---@return table|nil
function M.buildRecord(car, sim, state)
  if not state or type(state) ~= "table" then
    return nil
  end
  local lapsDone = tonumber(state.lapsCompleted) or 0
  if lapsDone <= 0 then
    return nil
  end

  local hist = state.lapFeatureHistory
  local lap_history, histBest, histLast, histAvg = lapHistorySummaries(hist)
  local bestMs = tonumber(state.bestLapMs) or histBest
  local lastMs = tonumber(state.lastLapMs) or histLast
  local avgMs = histAvg

  local lastCorners = {}
  if hist and type(hist) == "table" and #hist > 0 then
    local last = hist[#hist]
    if type(last) == "table" and type(last.corners) == "table" then
      lastCorners = simplifyCorners(last.corners)
    end
  end

  local grip = safeSimField(sim, "trackGripLevel")
  grip = tonumber(grip)

  local sk = (car and sim) and persistence.sessionKey(car, sim) or "unknown"

  return {
    schema_version = SCHEMA_VERSION,
    exported_at = isoUtcNow(),
    app_version_ui = state.appVersionUi or "",
    session_key = sk,
    car = {
      id = resolveCarId(car),
    },
    track = {
      id = resolveTrackId(sim),
    },
    conditions = {
      track_grip = grip,
    },
    summary = {
      laps_completed = lapsDone,
      best_lap_ms = bestMs,
      last_lap_ms = lastMs,
      avg_lap_ms = avgMs,
    },
    lap_history = lap_history,
    corners_last_lap = lastCorners,
    coaching_hints_last = serializeCoaching(state.coachingLines),
  }
end

--- Write journal JSON + append one line to `journal_index.jsonl`. No-op if no laps completed.
---@param car ac.StateCar|nil
---@param sim ac.StateSim|nil
---@param state table
---@return boolean
function M.writeSessionEnd(car, sim, state)
  local rec = M.buildRecord(car, sim, state)
  if not rec then
    return false
  end
  local skSafe = rec.session_key:gsub("[^%w%.%-_]+", "_")
  local fname = "session_" .. fileTimestampUtc() .. "_" .. skSafe .. ".json"
  local path = journalDir() .. "/" .. fname
  persistence.ensureParentDirForFile(path)
  local raw = persistence.encodeJson(rec)
  if not raw then
    return false
  end
  local f = io.open(path, "w")
  if not f then
    logJournal("failed to open journal file for write: " .. tostring(path))
    return false
  end
  if not f:write(raw) then
    logJournal("failed to write journal file: " .. tostring(path))
    f:close()
    return false
  end
  f:close()

  local idxLine = persistence.encodeJson({
    journal_file = fname,
    exported_at = rec.exported_at,
    session_key = rec.session_key,
  })
  if idxLine then
    persistence.ensureParentDirForFile(indexPath())
    local ip = indexPath()
    local af = io.open(ip, "a")
    if af then
      af:write(idxLine .. "\n")
      af:close()
    else
      logJournal("failed to open journal index for append: " .. tostring(ip))
    end
  end
  return true
end

return M
